src/Features/CSharp/Portable/CodeRefactorings/UseExplicitType/UseExplicitTypeCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,112 @@
+            }
+
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var declaration = GetDeclaration(root, textSpan);
+            Debug.Assert(declaration.IsKind(SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement, SyntaxKind.DeclarationExpression));
+
+            var declaredType = s_useExplicitTypeHelper.FindAnalyzableType(declaration, semanticModel, cancellationToken);
+            if (declaredType == null)
+            {
+                return;
+            }
+
+            if (declaredType.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            if (!declaredType.Span.IntersectsWith(textSpan.Start))
+            {
+                return;
+            }
+
+            var state = State.Generate(declaration, semanticModel, optionSet,
+                isVariableDeclarationContext: declaration.IsKind(SyntaxKind.VariableDeclaration), cancellationToken: cancellationToken);
+
+            // UseExplicitType analyzer/fixer already gives an action in this case
+            if (s_useExplicitTypeHelper.IsStylePreferred(semanticModel, optionSet, state, cancellationToken))
+            {
+                return;
+            }
+
+            Debug.Assert(state != null, "analyzing a declaration and state is null.");
+            if (!s_useExplicitTypeHelper.TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    CSharpFeaturesResources.Use_explicit_type,
+                    c => UpdateDocumentAsync(document, declaredType, c)));
+        }
+
+        private static SyntaxNode GetDeclaration(SyntaxNode root, TextSpan textSpan)
+        {
+            var token = root.FindToken(textSpan.Start);
+            return token.Parent?.FirstAncestorOrSelf<SyntaxNode>(
+                a => a.IsKind(SyntaxKind.DeclarationExpression, SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement));
+        }
+
+        private static async Task<Document> UpdateDocumentAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, document.Project.Solution.Workspace);
+
+            await UseExplicitTypeCodeFixProvider.HandleDeclarationAsync(document, editor, node, cancellationToken).ConfigureAwait(false);
```

Comment:
editor seems doesn't need to be created outside? are there reason HandlEdclearationAsync needs to accept editor which one can create from document given in? is it due to some sharing code?

Reply:
HandleDeclarationAsync can be called multiple times.  For example, when doing a fix-all.  So it needs the editor passed into it so it can make all the edits, instead of just making the editor itself and doing a single edit.